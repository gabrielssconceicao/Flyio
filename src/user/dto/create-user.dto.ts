import { ApiProperty } from '@nestjs/swagger';
import {
  IsEmail,
  IsNotEmpty,
  IsOptional,
  IsString,
  MaxLength,
  MinLength,
} from 'class-validator';
export class CreateUserDto {
  @ApiProperty({
    example: 'John Doe',
    description: 'User name',
    minLength: 3,
    maxLength: 255,
  })
  @IsString()
  @IsNotEmpty()
  @MinLength(3)
  @MaxLength(255)
  name: string;

  @ApiProperty({
    example: 'jdoe@email.com',
    description: 'User email',
  })
  @IsEmail()
  email: string;

  @ApiProperty({
    example: '123456',
    description: 'User password',
    minLength: 6,
    maxLength: 255,
  })
  @IsString()
  @IsNotEmpty()
  @MinLength(6)
  @MaxLength(255)
  password: string;

  @ApiProperty({
    example: 'This is my bio',
    description: 'User bio',
    maxLength: 255,
  })
  @IsOptional()
  @IsString()
  @MaxLength(255)
  bio?: string;

  @ApiProperty({
    example: 'example.png',
    description: 'User profile image',
  })
  @IsOptional()
  @IsString()
  profileImg?: string;
}
