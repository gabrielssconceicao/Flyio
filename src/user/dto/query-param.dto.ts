import { Type } from 'class-transformer';
import { IsOptional, IsString, Max, Min } from 'class-validator';

export class QueryParamDto {
  @IsOptional()
  @IsString()
  search?: string;
  @IsOptional()
  @Min(1)
  @Max(50)
  @Type(() => Number)
  limit?: number;
  @IsOptional()
  @Min(0)
  @Type(() => Number)
  offset?: number;
}
