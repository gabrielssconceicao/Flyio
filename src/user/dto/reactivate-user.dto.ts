import { ApiProperty } from '@nestjs/swagger';
import { IsNotEmpty, IsString } from 'class-validator';
import {
  dtoErrorMessages,
  DtoErrorType,
} from '../../common/utils/dto-error-messages';

export class ReactivateUserDto {
  @ApiProperty({
    example: 'krcio48hgGYIFYJfuyfJYAYu',
    description: 'Activation token',
  })
  @IsNotEmpty({ message: dtoErrorMessages('Token', DtoErrorType.REQUIRED) })
  @IsString()
  token: string;
}
