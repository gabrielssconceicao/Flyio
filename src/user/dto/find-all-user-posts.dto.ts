import { ApiProperty, OmitType } from '@nestjs/swagger';
import { FindAllResponseDto } from 'src/common/dto/find-all-response.dto';
import { PostEntity } from '../../post/entities/post.entity';

export class FindAllUsersPostsResponseDto extends FindAllResponseDto<
  Omit<PostEntity, 'user'>
> {
  @ApiProperty({
    type: [OmitType(PostEntity, ['user'])],
    description: 'List of posts of a user with limited details',
  })
  items: Omit<PostEntity, 'user'>[];
}
