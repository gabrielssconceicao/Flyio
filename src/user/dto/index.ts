import { CreateUserDto } from './create-user.dto';
import { FindAllUsersResponseDto } from './find-all-users.dto';
import { QueryParamDto } from './query-param.dto';
import { ReactivateUserDto } from './reactivate-user.dto';
import { UpdateUserDto } from './update-user.dto';

export {
  CreateUserDto,
  FindAllUsersResponseDto,
  QueryParamDto,
  ReactivateUserDto,
  UpdateUserDto,
};
