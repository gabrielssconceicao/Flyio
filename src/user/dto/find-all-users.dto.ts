import { ApiProperty } from '@nestjs/swagger';
import { PickType } from '@nestjs/mapped-types';
import { User } from '../entities/user.entity';
import { FindAllResponseDto } from '../../common/dto/find-all-response.dto';

class UserSummaryDto extends PickType(User, [
  'id',
  'name',
  'username',
  'profileImg',
] as const) {
  @ApiProperty({
    example: 'd-43-5df-df4',
    description: 'Unique identifier of the user',
  })
  id: string;

  @ApiProperty({ example: 'John Doe', description: 'Name of the user' })
  name: string;

  @ApiProperty({ example: 'jdoe', description: 'Username of the user' })
  username: string;

  @ApiProperty({
    example: 'https://example.com/profile.jpg',
    description: 'Profile image URL',
    nullable: true,
  })
  profileImg: string | null;
}

export class FindAllUsersResponseDto extends FindAllResponseDto<UserSummaryDto> {
  @ApiProperty({
    type: [UserSummaryDto],
    description: 'List of users with limited details',
  })
  items: UserSummaryDto[];
}
