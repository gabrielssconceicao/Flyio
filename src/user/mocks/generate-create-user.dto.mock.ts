import { CreateUserDto } from '../dto';
import { userMock } from './user.mock';
export const generateCreateUserDtoMock = (): CreateUserDto => {
  const dto: CreateUserDto = {
    name: userMock.name,
    username: userMock.username,
    email: userMock.email,
    password: userMock.password,
    bio: userMock.bio,
  };

  return dto;
};
