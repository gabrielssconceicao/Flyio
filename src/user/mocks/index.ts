import { generateUserMock } from './user.mock';
import { generateFindAllUsersResponseDtoMock } from './generate-find-all-users-response.dto.mock';
import { generateCreateUserDtoMock } from './generate-create-user.dto.mock';

export {
  generateUserMock,
  generateFindAllUsersResponseDtoMock,
  generateCreateUserDtoMock,
};
