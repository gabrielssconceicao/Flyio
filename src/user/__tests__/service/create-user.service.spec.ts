import { Test, TestingModule } from '@nestjs/testing';
import { JwtService } from '@nestjs/jwt';
import { ConflictException } from '@nestjs/common';
import { User } from '../../entities/user.entity';
import { generateCreateUserDtoMock, generateUserMock } from '../../mocks';
import { UserService } from '../../user.service';
import { CreateUserDto } from '../../dto';
import { HashingServiceProtocol } from 'src/auth/hashing/hashing.service';
import { hashingServiceMock, jwtServiceMock } from 'src/auth/mocks';
import { PrismaService } from 'src/prisma/prisma.service';
import { CloudinaryService } from 'src/cloudinary/cloudinary.service';
import {
  cloudinaryServiceMock,
  generatedProfilePictureMock,
  generateFileMock,
} from 'src/cloudinary/mocks';
import { PermissionService } from 'src/permission/permission.service';

describe('<UserService - (create) />', () => {
  let service: UserService;
  let hashingService: HashingServiceProtocol;
  let prismaService: PrismaService;
  let cloudinaryService: CloudinaryService;

  let hashedPassword: string;
  let user: User;
  let createUserDto: CreateUserDto;
  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        UserService,
        {
          provide: HashingServiceProtocol,
          useValue: hashingServiceMock,
        },
        {
          provide: PrismaService,
          useValue: {
            user: {
              create: jest.fn(),
              findFirst: jest.fn(),
            },
          },
        },
        {
          provide: CloudinaryService,
          useValue: cloudinaryServiceMock,
        },
        {
          provide: JwtService,
          useValue: jwtServiceMock,
        },
        {
          provide: PermissionService,
          useValue: {},
        },
      ],
    }).compile();

    service = module.get<UserService>(UserService);
    hashingService = module.get<HashingServiceProtocol>(HashingServiceProtocol);
    prismaService = module.get<PrismaService>(PrismaService);
    cloudinaryService = module.get<CloudinaryService>(CloudinaryService);

    hashedPassword = 'hashedPassword';
    user = generateUserMock();
    createUserDto = generateCreateUserDtoMock();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
    expect(hashingService).toBeDefined();
    expect(prismaService).toBeDefined();
    expect(cloudinaryService).toBeDefined();
  });

  it('should create a user without profile picture successfully', async () => {
    user.profileImg = null;
    jest.spyOn(prismaService.user, 'findFirst').mockResolvedValue(null);
    jest.spyOn(hashingService, 'hash').mockResolvedValue(hashedPassword);
    jest.spyOn(prismaService.user, 'create').mockResolvedValue({
      ...user,
      _count: { followers: 0, following: 0 },
    } as any);
    const result = await service.create(createUserDto, null);
    expect(result).toEqual(user);
    expect(result.profileImg).toBeNull();
    expect(prismaService.user.findFirst).toHaveBeenCalled();
    expect(cloudinaryService.uploadProfilePicture).not.toHaveBeenCalled();
    expect(hashingService.hash).toHaveBeenCalledWith(createUserDto.password);
    expect(prismaService.user.create).toHaveBeenCalled();
    expect(result).toMatchSnapshot();
  });

  it('should create a user with profile picture successfully', async () => {
    jest.spyOn(prismaService.user, 'findFirst').mockResolvedValue(null);
    jest
      .spyOn(cloudinaryService, 'uploadProfilePicture')
      .mockResolvedValue(generatedProfilePictureMock);
    jest.spyOn(hashingService, 'hash').mockResolvedValue(hashedPassword);
    jest.spyOn(prismaService.user, 'create').mockResolvedValue({
      ...user,
      _count: { followers: 0, following: 0 },
    } as any);

    const file = generateFileMock();
    const result = await service.create(createUserDto, file);
    expect(result).toEqual(user);
    expect(prismaService.user.findFirst).toHaveBeenCalled();
    expect(cloudinaryService.uploadProfilePicture).toHaveBeenCalledWith(file);
    expect(hashingService.hash).toHaveBeenCalledWith(createUserDto.password);
    expect(prismaService.user.create).toHaveBeenCalled();
    expect(result).toMatchSnapshot();
  });

  it('should throw ConflictException if user already exists', async () => {
    jest.spyOn(prismaService.user, 'findFirst').mockResolvedValue({} as any);

    await expect(service.create(createUserDto, null)).rejects.toThrow(
      ConflictException,
    );
  });
});
