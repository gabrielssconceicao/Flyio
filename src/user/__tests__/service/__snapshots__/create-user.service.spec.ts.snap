// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserService - (create) /> should create a user with profile picture successfully 1`] = `
{
  "active": true,
  "bio": "This is my bio",
  "followers": 0,
  "following": 0,
  "id": "42-d-f-df4",
  "name": "John Doe",
  "profileImg": "https://example.com/profile.jpg",
  "username": "jDoe453",
}
`;

exports[`<UserService - (create) /> should create a user without profile picture successfully 1`] = `
{
  "active": true,
  "bio": "This is my bio",
  "followers": 0,
  "following": 0,
  "id": "42-d-f-df4",
  "name": "John Doe",
  "profileImg": null,
  "username": "jDoe453",
}
`;
